[TOPIC-GPIO] validate pin ordinals in wrapper functions

pabigot opened the pull request:
Extend the driver data structure with a field that specifies the number of pins present on the device.  Document the fields and who is responsible for maintaining them.
    
Update all configuration functions for specific pins to return an error if the pin is not supported.
    
Update all set/get functions for specific pins to assert if the pin is not supported.

Fix const qualifier on driver data pointers.

Driver Update Status:

Driver | Updated |Run-tested
-- | -- | -- 
cc13xx_cc26xx | :heavy_check_mark: | 
cc2650 | :heavy_check_mark: | 
cc32xx | :heavy_check_mark: | 
cmsdk_ahb | :heavy_check_mark: | 
dw | :heavy_check_mark: | 
esp32 | :heavy_check_mark: | esp32
gecko | :white_check_mark: | efr32mg_sltb004a
ht16k33 | :heavy_check_mark: | 
imx | :white_check_mark: | 
intel_apl | :white_check_mark: | 
mchp_xec | :heavy_check_mark: | 
mcux | :heavy_check_mark: | frdm_k64f
mcux_igpio | :heavy_check_mark: | mimxrt1060_evk
mcux_lpc | :heavy_check_mark: | lpcxpresso55x69_cpu0 *requires rev A2 board*
mmio32 | :heavy_check_mark: | 
nrfx | :heavy_check_mark: | nrf52840_pca10056
pca95xx | :heavy_check_mark: | 
rv32m1 | :heavy_check_mark: | rv32m1_vega_ri5cy
sam | :heavy_check_mark: | sam_e70_xplained
sam0 | :heavy_check_mark: | atsamd21_xpro
sifive | :heavy_check_mark: | hifive1_revb *test hangs, timer issues?*
stellaris | :heavy_check_mark: | 
stm32 | :heavy_check_mark: | nucleo_l476rg
sx1509b | :heavy_check_mark: | nrf52_pca20020

* :heavy_check_mark: probably good
* :white_check_mark: see commit message for limitations

Background: This implements a discussion in slack `#gpio` 2019-10-09T16:04Z quoted below:

> OK, so before doing pin validation in the wrapper functions we need to agree on what's being validated.  "Simple Poll" doesn't seem to work in this channel, so up-vote the options you want.
>* A pin index that is not less than the number of pins should cause -EINVAL to be returned *received 3+2=5 thumbs up*
>* A (port) mask that has bits set that correspond to non-existing pins should cause -EINVAL to be returned *received no votes*
>* A (port) value that has bits set that correspond to non-existing pins should cause -EINVAL to be returned *received no votes*
>
> Consider that it may be useful to pass -1 as a value or mask to indicate all pins, which is why I'm voting only for the first.

pabigot committed:
drivers: gpio: fix const qualifier on driver data pointers

Several declarations provided a const pointer to mutable data.  Change
all declarations that are not used to change the invert field to be
pointers to const data.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

galak commented:
This works well with the recently added DTS property `ngpios`.

pabigot committed:
drivers: gpio: add support for device-specific pin validation

Extend the driver data structure with a field that identifies the pins
supported by the device.  Document the fields and who is responsible
for maintaining them.

Update all configuration functions for specific pins to return an
error if the pin is not supported.

Update all set/get functions for specific pins to assert if the pin is
not supported.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Explicitly marked DNM: between two commits in this every driver needs to be updated to set the ngpios field.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -405,7 +405,18 @@ typedef u32_t gpio_port_value_t;
+	 *
+	 * This must be set by the driver during device initialization.
+	 */
+	u8_t num_pins;
```
Number of pins can be determined during the compile time and is constant. We could put it inside `dev->config->config_info` structure? That saves 4 bytes of RAM. It is however more work since we currently don't expose `dev->config->config_info` at the API level.

pabigot replied:
We'd also have to define a common structure that everybody would have to put at the start of their private config structure.   Since we already do that for data, that's not too far a stretch, but it touches every driver.

It also can't be done until the work started in #20034 is [completed](https://github.com/zephyrproject-rtos/zephyr/pull/20034#discussion_r338619559).

pabigot replied:
Moved to common config structure, some drivers updated.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -602,8 +602,8 @@ static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
 		(const struct gpio_driver_api *)port->driver_api;
-	struct gpio_driver_data *const data =
-		(struct gpio_driver_data *const)port->driver_data;
+	const struct gpio_driver_data *data =
```
A constant pointer to mutable data is what we want. The pattern is common in the driver code, typically encoded by `DEV_DATA` macro. We could possibly do
```
const struct gpio_driver_data *const data =
```
but I don't think it's worth the trouble. We don't do it inside the drivers even if the function accesses `driver_data` read only.

pabigot replied:
> A constant pointer to mutable data is what we want.

Well, no, it's what you want.  I think it's bizarre and unjustified, though I do see that some (not all) `DEV_DATA` macros create a const value referencing mutable data.  I'm having difficulty understanding why; the only value I can is it would stop somebody from doing:

    DEV_DATA(drv) = some_value;

I don't normally make the pointers const because people get confused and it doesn't really matter if somebody re-uses the variable with a different value, but I don't object to putting that back.  I didn't do it here because most driver adaptation headers don't do it, including other uses in this one.

I certainly do create pointers to `const` objects when I have no intention of modifying the object contents.  It lets the maintainer understand my intent.  

Are you absolutely sure you want const pointers to mutable data when the data will not be changed?  If so, I'll drop that cleanup commit and let you make the header consistent.

pabigot replied:
Used `const` qualifier on variable declaration.

mnkp reviewed (changes requested):
@pabigot thanks for the PR and sorry for a long delay reviewing it.

mnkp commented:
@pabigot Let's discuss open points in the API meeting.

carlescufi commented:
@pabigot as per API meeting, there is a strong preference for saving memory in RAM, meaning that the `num_pins` belongs in the `config` structure instead.

pabigot committed:
dts: binding: gpio: add default for ngpios property

Almost all drivers support 32 pins, though some support less.
Eventually this should be specified in every SOC binding, but until
then provide a default so GPIO drivers can be updated to use this.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_cc13xx_cc26xx: add support for pin validation

Add a config structure and use the devicetree GPIO pin count to
initialize it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_cc32xx: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_cmsdk_ahb: add support for pin validation

Replace the common data structure with the common config structure as
a prefix of the driver-specific config structure and use the
devicetree GPIO pin counts to initialize it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_dw: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sx1509b: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_nrfx: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
I've updated this with support for storing the number of pins in the static config structure, and have updated several drivers.  I've tested nrfx and sx1509b.  Note that once merged any drivers that have not been updated will fail because they'll be misinterpreting the config structure.

I'm willing to continue converting the drivers so they're all in this PR but I won't be able to test many of them, and I'll be using the `DT_INST_#_COMPAT_NGPIOS` spelling of the devicetree property.  So there may be per-driver fixups needed later.

Please provide any feedback on the general approach and I'll resume updating drivers later this week.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1597,16 +1597,11 @@ def _check_prop_type_and_default(prop_name, prop_type, required, default,
-    if required:
-        _err("'default:' for '{}' in 'properties:' in {} is meaningless in "
-             "combination with 'required: true'"
-             .format(prop_name, binding_path))
```
Might be better to just remove this check and still let the checks below run when there's a default.

pabigot replied:
Could you explain that?  If the property is required the default should be superseded by the value provided in the devicetree source, so there's no point continuing to validate the default.  In particular, it might not be compatible with the type of the property if that changed when the property was made `required` in an override.

ulfalizer replied:
Nice to flag stuff like this, as it probably wasn't intended:

```
foo:
    required: true
    type: int
    default: "cool string that doesn't make sense"
```

pabigot replied:
It would be, but I don't know how to do that, and I'm not going to fuss with it.

I've opened #21181 for this.  This PR will be blocked until that's resolved.  If the decision is that we have to update the SOC bindings for every target platform, then that's what we'll do.

ulfalizer replied:
Would just need to do this I think, removing all other changes:

```diff
diff --git a/scripts/dts/edtlib.py b/scripts/dts/edtlib.py
index 9877f1157e..84532fca54 100644
--- a/scripts/dts/edtlib.py
+++ b/scripts/dts/edtlib.py
@@ -1604,11 +1604,6 @@ def _check_prop_type_and_default(prop_name, prop_type, required, default,
     if default is None:
         return
 
-    if required:
-        _err("'default:' for '{}' in 'properties:' in {} is meaningless in "
-             "combination with 'required: true'"
-             .format(prop_name, binding_path))
-
     if prop_type in {"boolean", "compound", "phandle", "phandles",
                      "phandle-array"}:
         _err("'default:' can't be combined with 'type: {}' for '{}' in "
```

pabigot replied:
Ah; sorry, I thought you were giving an example where the property type changed in an override, which was a case I was concerned about.  OK, I've just removed the check completely.

pabigot commented:
Blocked by #21181.

ulfalizer reviewed (approved):
Now checks `default:` against `type:` for `required: true`.

karstenkoenig commented:
Sorry I just stumbled upon this due to the discussion in #gpio
Is the main purpose of this PR to check if a pin is valid? Because just comparing against the number of pins on the port is often not enough, at least in certain package variants there might be gaps. An example would be the STM32L053 (pretty random pick from historic memory), in the 64pin packages it has PC0-PC15, where as the 48pin package only has PC13-PC15. Wouldn't a bitmask of valid pins work just as well and consume just as much space? It'd be more expensive to check for but would be more flexible.

pabigot commented:
I think the purpose is indeed to detect and reject invalid pins and reject them.  I wasn't aware of controllers that had gaps; for those we do need something like a bitmask.  @galak introduced `ngpios`, maybe it wasn't the right solution.

In any case, from my perspective the whole PR should be stalled until we figure out what the right solution is.  At this point every driver is doing its own validation internally (or not validating at all); maybe we should stick with that.

karstenkoenig commented:
I believe a pure range based check will not be generic enough to deliver the promise of validating if a pin is existing. A bitmask might pull it off, I can't think of a case where it wouldn't, but maintaining bitmasks by package certainly ain't great and not sure if a 32bit mask would work for all platforms though.
We had a similar situation on the CAN api where drivers performed different levels of input sanitation for an invalid frame length, creating the situation that one can not rely on input validation at all. It got resolved by the test being implemented everywhere, but it's not really mandated by the interface description. A discussion about input sanitation might make sense for Zephyr in general, because input sanitation can be a bottomless hole, yet for could be argued for in a functional safety aimed RTOS.

erwango commented:
I do agree with @karstenkoenig a bitmap would be the right thing to do here.
Though providing the information is not an easy thing to do (at least on STM32 side) since it would consist in adding a new level in STM32 dtsi hierarchy (maybe we can think of something smarter, but in any case this requires careful and complete study).
Besides, since information is not there today, validation cannot be done in a better way inside the driver. Last, this proposal, even if not perfect is better than current solution anyway.

So I'd propose to keep this improvement for a next evolution, once we're ready to provide full package information.

carlescufi commented:
As per the API meeting the general direction of this PR is acknowledged as correct, and support for reserved ranges may be added later on top of the functionality provided here.

pabigot commented:
Based on the API meeting yesterday, and the subsequent introduction of #21277, I'm going to proceed under the assumption that eventually something will be producing a bitmask of valid pins, and will change the config value and the verification to be based on that interface.  For now I'll initialize the mask to `BIT(ngpios) - 1U`, but presumably eventually the `gpio-ranges` pairs will be translated into something that produces that bitmask for the controller and exposes it as a generated macro value.

pabigot committed:
drivers: gpio_esp32: add support for pin validation

Add a config structure for each port and use the devicetree GPIO pin
count to initialize it.  Simplify device initialization by using
instance number as only variation point.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_gecko: add hard-coded support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure.  This driver uses fixups to define a common name
based on the SOC; until aliases for the new properties are added the
code assumes all pins are supported.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_ht16k33: add hard-coded support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure.  The devicetree binding does not inherit from
gpio-controller, so the required property is not available there.  A
hard-coded value of 32 pins has been implemented to make the driver
function.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_imx: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.  Not all aliases appear to be present, but using instances breaks
because the ports used by different boards do not always start with
the first.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_intel_apl: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the hard-coded device-specific pin count to
initialize it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_mchp_xec: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_mcux: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_mcux_igpio: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_mcux_lpc: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_mmio32: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the mask parameter to initialize it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_pca95xx: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_rv32m1: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sam: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sam0: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sifive: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_stellaris: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure and use the devicetree GPIO pin counts to initialize
it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_stm32: add support for pin validation

Add the common config structure as a prefix of the driver-specific
config structure.  This driver uses fixups to define a common name
based on the SOC series; until aliases for the new properties are
added the code assumes all pins are supported.

Since pin ordinal validation is now done directly in gpio API,
remove checks inside the driver itself.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -627,6 +627,9 @@ static int gpio_stm32_init(struct device *device)
 #define GPIO_DEVICE_INIT(__name, __suffix, __base_addr, __port, __cenr, __bus) \
 	static const struct gpio_stm32_config gpio_stm32_cfg_## __suffix = {   \
+		.common = {						       \
+			.available = (gpio_port_pins_t)-1,		       \
```
I actually spent some time to understand how substracting 1 to gpio_port_pins_t could work. Anyway to disambiguate ?

Then, STM32 GPIO ports actually control 16 bits, so we should use `GPIO_AVAILABLE_PINS` macro, checking ngpios value. Bright side is that code will be ready for update to a range version.
Adding the whole fixup code is a bit cumbersome, so I'm preparing a commit to spare you the effort.

Last, I've tried to set `ngpios` by default to 16 `st,stm32-gpio.yaml`, but it is actually not possible to override the 32 value defined in `gpio-controller`. Is there a strong requirement to prevent this?
I'm not quite in favor of spreading dts values in bindings, but this is indeed handy for this kind of things.

pabigot replied:
I suppose `(gpio_port_pins_t)(-1)` might provide a clue where necessary.

If the GPIO ports control 16 bits, then this should be `GPIO_AVAILABLE_PINS(16U)`.  I could make that change.  IIRC I went with the literal because the implementation depends on fixups to get the right devicetree value and I didn't want to add all of those (I ultimately had to for IMX because reasons, but it's not good).  I'd welcome a patch that turns it into what you want.

I don't see any reason not to allow overriding an existing default; the whole dts binding description architecture is pretty ad hoc.  This PR requires and includes #21271; if that doesn't get merged soon I'll need to push for it.  If you make a PR to support default override we can include it here too.

erwango replied:
+1 for `(gpio_port_pins_t)(-1)`
`GPIO_AVAILABLE_PINS(16U)` is fine with me as well. I can push an update later on to work with a dts extracted info, but all in all, I prefer this than the bindings inferred default `ngpios` value.

pabigot replied:
OK, I'll queue those changes in the next update.

erwango replied:
Thanks!

pabigot replied:
Done.

erwango reviewed (approved):
Tested ok

pabigot committed:
tests: gpio_basic_api: add overlay for hifive1_revb testing

Use the Arduino A4 and A5 pins which have no dedicated purpose.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

erwango committed:
drivers/gpio: stm32: Align arguments types on few api functions

Somehow, types used in some api functions where not in line with
api.
Fix this.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

henrikbrixandersen reviewed (approved):
HT16K33 changes OK.

pabigot commented:
@mnkp The issues in your previous review were addressed after the meeting where we discussed them.  Please re-check.

mnkp reviewed (commented):
@pabigot I believe I misunderstood your comment during the API meeting. I thought the idea was to do a rough check based on ngpios property in the API and leave a more detailed check to be implemented by the driver. The current solution with bit mask is very convenient and flexible. However, very likely it will not be future proof. And looking into the future I believe we need to do a better job to support hardware GPIO modules like the Intel one #20400 where one GPIO controller can have up to 78 pins.

To improve support for such GPIO modules we would need to add _pin_ functions to the driver API (the user GPIO API wouldn't change). In this case having check based on mask wouldn't work. We would need a solution like in Linux where they define list of valid pin ranges.

pabigot commented:
https://github.com/zephyrproject-rtos/zephyr/pull/20115#discussion_r361815226 changed to assert.

For the overall approach:  https://github.com/zephyrproject-rtos/zephyr/pull/20115#issuecomment-564560462 explains the switch to bitmask.  Since the intent is to get `gpio-reserved-ranges` in soon this is the future-proof path that allows drivers that need the feature to use it.

Given the difficulty in getting this GPIO API change in (over 8 months so far since #15611), I can't see that another rework to support more than 32 pins per gpio controller will ever happen.  Nexus maps mostly allow the limitation to be hidden from developer code, as demonstrated for the up-squared platform.

mnkp reviewed (commented):
> Nexus maps mostly allow the limitation to be hidden from developer code, as demonstrated for the up-squared platform.

This is a good workaround but it's still a workaround. Adding pin level functions to GPIO drivers will be significantly less work than what we are doing now. It's not going to happen tomorrow, there are more important things to work on but at some point we should to it.

We can also change pin validation code later (I have some more comments to the current implementation in this case). Indeed, we shouldn't delay the current work.

There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.

pabigot commented on drivers/gpio/gpio_utils.h:
```diff
@@ -11,6 +11,9 @@
 
+#define GPIO_AVAILABLE_PINS(ngpios) ((ngpios == 32U)		\
+				     ? ((gpio_port_pins_t)(-1))	\
+				     : (BIT(ngpios) - 1U))
```
No, it has to be this way.  `BIT(32)` produces an overflow when `1UL << 32` is evaluated and unsigned long is a 32-bit value.

mnkp replied:
I misread the purpose of the code. But in this case it would be cleaner to cast to long long or uintmax_t. This will be computed during the compilation time, shouldn't influence the code size.

pabigot commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -627,6 +627,9 @@ static int gpio_stm32_init(struct device *device)
 #define GPIO_DEVICE_INIT(__name, __suffix, __base_addr, __port, __cenr, __bus) \
 	static const struct gpio_stm32_config gpio_stm32_cfg_## __suffix = {   \
+		.common = {						       \
+			 .available = GPIO_AVAILABLE_PINS(16U),		       \
```
See the commit message: it's not possible to do that here, and I believe @erwango approved this as the short-term workaround.

pabigot commented on drivers/gpio/gpio_utils.h:
```diff
@@ -11,6 +11,9 @@
 #ifndef ZEPHYR_DRIVERS_GPIO_GPIO_UTILS_H_
 #define ZEPHYR_DRIVERS_GPIO_GPIO_UTILS_H_
 
+#define GPIO_AVAILABLE_PINS(ngpios) ((ngpios == 32U)		\
```
I can do `GPIO_PORT_PINS_MASK_FROM_NGPIOS` here, and change `available` to `port_pins_mask`.

pabigot replied:
Should all be updated.

pabigot commented:
> There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.

That's going to have to be the responsibility of the folks doing the devicetree scripting support.  There'll have to be custom code that generates the appropriate mask for all identified pins.

mnkp commented:
> > There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.
> 
> That's going to have to be the responsibility of the folks doing the devicetree scripting support. There'll have to be custom code that generates the appropriate mask for all identified pins.

In this case, shouldn't we generate the pin mask already, without support for `gpio-reserved-ranges`, or otherwise we will need to touch all the gpio drivers again. Also, we would make sure that this custom code is fine with devicetree folks.

pabigot commented:
> In this case, shouldn't we generate the pin mask already, without support for `gpio-reserved-ranges`, or otherwise we will need to touch all the gpio drivers again. Also, we would make sure that this custom code is fine with devicetree folks.

Only drivers for controllers that have gaps in the supported pins would have to be updated: every other driver can use the mask-from-ngpios code this PR provides.  I don't want to hold this up for another release because we're stalled waiting for more devicetree tooling to get implemented.

mnkp commented:
> Only drivers for controllers that have gaps in the supported pins would have to be updated: every other driver can use the mask-from-ngpios code this PR provides.

If adding the custom code to DTS scripts to generate the mask is not a problem, then that's indeed a good approach.

mnkp commented on include/drivers/gpio/gpio_mmio32.h:
```diff
@@ -44,7 +46,10 @@ int gpio_mmio32_init(struct device *dev);
 static const struct gpio_mmio32_config _dev_name##_dev_cfg = {		\
-	.reg	= (volatile u32_t *)_address,			\
+	.common = {							\
+		 .available = _mask,					\
```
`.port_pin_mask`

mnkp commented on include/drivers/gpio.h:
```diff
@@ -996,14 +1035,18 @@ static inline int gpio_port_set_clr_bits(struct device *port,
-	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");
+	(void)cfg;
+	__ASSERT((cfg->port_pin_mask & (gpio_port_pins_t)BIT(pin)) != 0U,
+		 "Invalid pin number");
```
Other asserts print "Unsupported pin", we could use a standard message.

mnkp commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -124,6 +124,38 @@ static int setup(void)
 }
 
+/* Verify pin-number validation on all non-deprecated API that does
+ * this without __ASSERT.
```
All the functions use ASSERT now, the testcase can be removed (or we should write a testcase that tests the ASSERT).

pabigot replied:
Then that's wrong: the plan was to return `-EINVAL` when an invalid pin was provided to a function, per the decision recorded in the PR description.  We should either revert some of those changes to `ASSERT` or get agreement that it's OK to use only assert.  To discuss next API meeting.

pabigot replied:
Per API meeting: test will be removed.

pabigot replied:
done

pabigot committed:
x86: apollo_lake: correct NW GPIO controller offset and span

The offset and the number of pins were swapped causing an error when a
64-position left shift was generated to calculate the pin mask.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
